apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -104,122 +243,699 @@ terminate(_Reason, _St) ->
+                false ->
+                    lager:debug("status hasn't changed", []),
+                    {Status, C1}
+            end;
+        Other ->
+            lager:debug("Other = ~p", [Other]),
+            Other
+    end.
+
+handle_ch_status_changed(ChId, #{ last_block := Last
+                                , top_hash := Hash } = C) ->
+    lager:debug("channel status changed, Hash=~p, Last=~p", [Hash,Last]),
+    {TxLog, C1} = get_txs_since({any_after_block, Last}, Hash, ChId, C),
+    lager:debug("New TxLog: ~p", [TxLog]),
+    #{tx_hash := Tx, block_hash := BlockHash} = get_latest_tx(TxLog),
+    lager:debug("Latest Tx = ~p, Hash = ~p", [Tx, BlockHash]),
+    {Status, C2} = get_basic_ch_status_(ChId, BlockHash, C1),
+    lager:debug("ChStatus(~p) = ~p", [BlockHash, Status]),
+    %% Assert that this is really a changed channel object
+    true = channel_status_changed(maps:get(vsn, Status), C2),
+    lager:debug("asserted status changed", []),
+    Status1 = Status#{tx => Tx},
+    {Status1, C2#{{channel, BlockHash} => Status1}}.
+
+get_basic_ch_status_(ChId, #{ top_hash := Hash } = C) ->
+    get_basic_ch_status_(ChId, Hash, C).
+
+get_basic_ch_status_(ChId, BlockHash, C) ->
+    {Ch, C1} = get_channel(ChId, BlockHash, C),
+    case Ch of
+        undefined ->
+            {undefined, C1};
+        _ ->
+            Vsn = channel_vsn(Ch),
+            Changed = channel_status_changed(Vsn, C1),
+            {#{ is_active    => aesc_channels:is_active(Ch)
+              , vsn          => Vsn
+              , changed      => Changed
+              , channel      => Ch
+              , lock_period  => aesc_channels:lock_period(Ch)
+              , locked_until => aesc_channels:locked_until(Ch) },
+             C1}
+    end.
+
+get_channel(ChId, Hash, C) ->
+    cached_get({channel, Hash}, C, fun() -> get_channel_(ChId, Hash) end).
+
+get_channel_(ChId, Hash) ->
+    case aec_chain:get_block_state(Hash) of
+        {ok, Trees} ->
+            case aec_chain:get_channel(ChId, Trees) of
+                {ok, Ch} ->
+                    Ch;
+                {error, _} ->
+                    undefined
+            end;
+        error ->
+            undefined
+    end.
+
+cached_get(Key, C, Get) ->
+    case maps:find(Key, C) of
+        {ok, V} ->
+            {V, C};
+        error ->
+            get_and_cache(Get, Key, C)
+    end.
+
+get_and_cache(F, Key, C) when is_function(F, 0) ->
+    V = F(),
+    {V, C#{ Key => V }};
+get_and_cache(F, Key, C) when is_function(F, 1) ->
+    {V, C1} = F(C),
+    {V, C1#{ Key => V }}.
+
+log_tx(#{ tx_hash      := TxHash
+        , block_hash   := BlockHash
+        , block_origin := _Origin
+        , type         := _Type } = Info, #st{tx_log = TxLog} = St) ->
+    Key = {TxHash, BlockHash},
+    case aesc_window:keymember(Key, 1, TxLog) of
+        true ->
+            St;
+        false ->
+            TxLog1 = aesc_window:add({Key, Info}, TxLog),
```

Comment:
The second element in this tuple that is the entry of the `aesc_window` data structure has a `type` key here (see pattern in function clause head).

Reply:
I noticed this via stricter type checks in https://github.com/aeternity/aeternity/pull/2488

From https://circleci.com/gh/aeternity/aeternity/40229
```
apps/aechannel/src/aesc_fsm_min_depth_watcher.erl
 775: Record construction #tx_log_entry{key::{_,_},value::#{'block_hash':=_, 'block_origin':=_, 'tx_hash':=_, 'type':=_, _=>_}} violates the declared type of field value::#{'block_hash':=binary(), 'block_origin':='chain', 'tx_hash':=binary()}
```